### PR TITLE
[Sparse] Add SpSpMM for sparse-sparse and sparse-diag matrix multiplication.

### LINKED ISSUE #4857: [Sparse] Support sparse-sparse and sparse-diag matmul
## 🔨Work Item

**IMPORTANT:**
* This template is only for dev team to track project progress. For feature request or bug report, please use the corresponding issue templates.
* DO NOT create a new work item if the purpose is to fix an existing issue or feature request. We will directly use the issue in the project tracker.

Project tracker: https://github.com/orgs/dmlc/projects/2

### PR DESCRIPTION
**Pull request by czkkkkkk**
## Description
Resolve #4857.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [ ] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [ ] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: dgl_sparse/include/sparse/spspmm.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/**
```
Comment (frozenbugs) [inline dgl_sparse/include/sparse/spspmm.h:1]: It is probably a little bit overkill to create one header per method. In the future, the customers need to include multiple header files to do one thing.

Maybe merge them all to mm.h?

File: dgl_sparse/src/spspmm.cc [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        SpSpMMNoAutoGrad(lhs_mat->Transpose(), lhs_val, ret_mat, output_grad);
+    rhs_val_grad = _CSRMask(rhs_mat_grad, rhs_mat_grad->value(), rhs_mat);
+  }
+  return {torch::Tensor(), lhs_val_grad, torch::Tensor(), rhs_val_grad};
+}
+
+c10::intrusive_ptr<SparseMatrix> SpSpMM(
+    const c10::intrusive_ptr<SparseMatrix>& lhs_mat,
+    const c10::intrusive_ptr<SparseMatrix>& rhs_mat) {
+  auto results = SpSpMMAutoGrad::apply(
```
Comment (frozenbugs) [inline dgl_sparse/src/spspmm.cc:115]: Should _SpSpMMSanityCheck be called here?

File: python/dgl/mock_sparse2/matmul.py [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    >>> A2 = torch.randn(3, 2)
+    >>> result = A1 @ A2
+    >>> print(type(result))
+    <class 'torch.Tensor'>
+    >>> print(result.shape)
+    torch.Size([3, 2])
+    """
+    assert isinstance(
+        A2, (torch.Tensor, SparseMatrix, DiagMatrix)
+    ), f"Expect arg2 to be a torch Tensor, SparseMatrix, or DiagMatrix object, got {type(A2)}"
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/matmul.py:225]: 80 chars.

File: python/dgl/mock_sparse2/matmul.py [hunk 4]
```diff
@@ -53,11 +53,94 @@ def spmm(A: Union[SparseMatrix, DiagMatrix], X: torch.Tensor) -> torch.Tensor:
...
+    assert isinstance(
+        A1, (SparseMatrix, DiagMatrix)
+    ), f"Expect A1 to be a SparseMatrix or DiagMatrix object, got {type(A1)}"
+    assert isinstance(
+        A2, (SparseMatrix, DiagMatrix)
+    ), f"Expect A2 to be a SparseMatrix or DiagMatrix object, got {type(A2)}"
+
+    if isinstance(A1, DiagMatrix) and isinstance(A2, DiagMatrix):
+        return _diag_diag_mm(A1, A2)
+    if isinstance(A1, DiagMatrix):
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/matmul.py:129]: Will the performance be improved if we have dedicated Diag @ Sparse implementation?
If so, do we plan to revisit the implementation here?

File: python/dgl/mock_sparse2/matmul.py [hunk 5]
```diff
@@ -97,7 +180,54 @@ def mm_sp(
     if isinstance(A2, torch.Tensor):
         return spmm(A1, A2)
     else:
-        raise NotImplementedError
+        return spspmm(A1, A2)
+
+
+def mm_diag(
```
Comment (frozenbugs) [inline python/dgl/mock_sparse2/matmul.py:186]: Do we need to expose mm_diag and mm_sp? Will it better to just expose mm?

File: python/dgl/mock_sparse2/matmul.py (hunk 4 above)
Comment (czkkkkkk) [reply python/dgl/mock_sparse2/matmul.py:129]: It is possible, but it requires non-trivial implementation. Since Diag @ sparse is not frequently used, its performance improvement should not have high priority.

**Review by frozenbugs — APPROVED** [review]

File: python/dgl/mock_sparse2/sddmm.py [hunk 6]
```diff
@@ -20,14 +20,17 @@ def sddmm(
...
     Parameters
     ----------
     A : SparseMatrix
         Sparse matrix of shape `(M, N)`.
     mat1 : Tensor
-        Dense matrix of shape `(M, K)`
+        Dense matrix of shape `(M, K)` or `(M,)`
     mat2 : Tensor
-        Dense matrix of shape `(K, N)`
+        Dense matrix of shape `(K, N)` or `(K,)`
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/sddmm.py:33]:
```suggestion
        Dense matrix of shape `(K, N)` or `(N,)`
```

File: python/dgl/mock_sparse2/sddmm.py [hunk 7]
```diff
@@ -20,14 +20,17 @@ def sddmm(
     .. math::
         out = (mat1 @ mat2) * A
 
+    In particular, :attr:``mat1`` and :attr:``mat2`` can be 1-D, which will be
+    reshape as `(M, 1)` and `(1, K)` in the computation.
```
Comment (jermainewang) [inline python/dgl/mock_sparse2/sddmm.py:24]:
```
In particular, :attr:``mat1`` and :attr:``mat2`` can be 1-D, then ``mat1 @ mat2`` becomes
the out-product of the two vector (which results in a matrix).
```

File: tests/pytorch/mock_sparse2/test_matmul.py [hunk 8]
```diff
@@ -68,6 +68,17 @@ def test_sparse_sparse_mm(
     A1 = create_func1(shape1, nnz1, dev)
     A2 = create_func2(shape2, nnz2, dev)
 
+    print("------------A1------------")
```
Comment (frozenbugs) [inline tests/pytorch/mock_sparse2/test_matmul.py:71]: We don't need to print info in test.

File: dgl_sparse/src/spspmm.cc [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+ public:
+  static variable_list forward(
+      AutogradContext* ctx, c10::intrusive_ptr<SparseMatrix> lhs_mat,
+      torch::Tensor lhs_val, c10::intrusive_ptr<SparseMatrix> rhs_mat,
+      torch::Tensor rhs_val);
+
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
+
+void _SpSpMMSanityCheck(
```
Comment (frozenbugs) [inline dgl_sparse/src/spspmm.cc:30]: Should we check non-dup edges here for spspmm? or you plan to do it in a separate PR?

**Review by frozenbugs — APPROVED** [review]

File: dgl_sparse/src/spspmm.cc (hunk 9 above)
Comment (jermainewang) [reply dgl_sparse/src/spspmm.cc:30]: Will do it in another PR.
